First stab at cross-compilation support

roblabla opened the pull request:
My first stab at fixing #110. Not fully complete yet (need to integrate it into the CLI, write some tests, fix some others).

This PR works by adding a new optional target flag to the Create step, which contains the rust target to build for. This flag is passed as-is to the cargo invocation.

When invoking wix, we use the target to find which `-arch` to pass to wix. This is currently implemented by looking at the first component of the target name, and transforming it to something wix-compatible (e.g. "x86_64" => "x64", etc...). Note that this is somewhat wrong: Ideally we should extract the target spec from `rustc -Z unstable-options --print target-spec-json | jq .llvm-target`, but the target spec json isn't stable, so we can't easily rely on it. Using the target name is an OK approximation that should get us most of the way there.

We then pass the target name to Wix as a variable, so we can find where the resulting binaries go.

I took the opportunity to remove use of Win64 and Platform in the wxs: those attributes are deprecated according to the wix documentation, passing -arch on the command line should be prefered (which this PR now does). `var.Platform` is replaced by `sys.BUILDARCH`.

volks73 commented:
>I took the opportunity to remove use of Win64 and Platform in the wxs: those attributes are deprecated according to the wix documentation, passing -arch on the command line should be prefered (which this PR now does). 

I just want to note that the `Platform` attribute for the [Package](https://wixtoolset.org/documentation/manual/v3/xsd/wix/package.html) element is not deprecated but discouraged according to the online documentation. There is a some confusion as some values for the `Platform` attribute have been deprecated. I am guessing the discouraged usage of the `Platform` attribute is to eventually deprecate it. I am all for removing it. I just wanted to make a note to avoid future confusion.

However, I cannot find any information on the deprecation of the `Win64` attribute for the [Component](https://wixtoolset.org/documentation/manual/v3/xsd/wix/component.html) element. It appears to be automatically set based on the `-arch` switch, similar to the recommended usage for the `Platform` attribute. So, it looks like all that is really needed is to use the `-arch` switch and let WiX take care of everything else. This is all good. Again, just wanting to make notes to reduce confusion in the future.

volks73 commented:
By the way, the approach looks good. Let me know when you are ready to merge.

I agree with implementing a workaround until target-spec-json is stabilized. Let's just create an issue after this merged to remind us to refactor when appropriate.

roblabla commented:
This took longer than expected to finish, sorry about the delay :sweat_smile:. This is now ready for review. I mainly fixed all the tests and added a test to make sure cross-compilation actually worked, and added back `var.Platform` to avoid breaking old wxs (Its use should be avoided in the future though).

Big caveat: I have no clue about the good practices in WiX. Notably the whole CargoTargetExplicit ifdef thing feels gross to me. You can't exactly pass booleans to the preprocessor, so that was the next closest thing I could find, but if there's a better way, I'd love to know :).

volks73 commented:
>This took longer than expected to finish, sorry about the delay 😅

No worries. It might take me some time to review, too.

>...and added back `var.Platform` to avoid breaking old wxs (Its use should be avoided in the future though).

Without any in-depth review or extensive thought to this, I am actually okay with breaking compatibility here. If the usage should be avoided, then we just need eliminate it. I was going to make a v0.4 release after this was complete. The tick up in the minor version, pre-1.0.0, would indicate some things have changed and I would make note in the release notes. I will play around with it a little bit to see if my initial thoughts are constant.

roblabla commented:
> I was going to make a v0.4 release after this was complete. The tick up in the minor version, pre-1.0.0, would indicate some things have changed and I would make note in the release notes.

My thought process is, since there's no (easy) way to tell cargo install "Please install this version or any semver-compatible version", I tend to avoid breaking changes in tools installable via `cargo-install` as much as possible. I have some internal tooling in my CI to pin my tools to specific version, but I assume this isn't the case of most people.

So removing the `var.Platform` variable will likely break a lot of people's build, while keeping it only costs us a single extra line of code. So I think keeping it around is worth the trouble.

volks73 commented:
I just looked at the `main.wxs.mustache` template, and the `var.Platform` is removed. Maybe the commit to add it back in was not pushed?

roblabla commented:
@volks73 the template doesn't use it anymore, but it's still defined in `create.rs` such that people using the old template can still build.

volks73 commented:
> The template doesn't use it anymore, but it's still defined in `create.rs` such that people using the old template (e.g. people who did a `cargo wix init` with the previous version) can still build.

Got it! The platform is now just an alias for the architecture.

volks73 commented on src/templates/main.wxs.mustache:
```diff
@@ -21,12 +21,16 @@
+    <?define PlatformProgramFilesFolder = "ProgramFilesFolder" ?>
+<?endif ?>
+
+<?ifdef CargoTargetExplicit ?>
+    <?define CargoTargetPathName = "$(var.CargoTarget)" ?>
+<?else ?>
+    <?define CargoTargetPathName = "" ?>
```
I do wonder, can we just collapse the `CargoTarget`, `CargoTargetPathName`, and `Profile` variables down into a single variable? Then, in the cargo-wix _create_ implementation, we build the "full" path to the binary? I think this would avoid needing the `CargoTargetExplicit` hack. We don't allow overriding the `Cargo` variables in the CLI...yet. 

Partly, I am just really confused by all of this, which is probably due to my lack of understanding about the cross compilation process with Cargo. So, please humor me with some external thoughts and ramblings:

At the moment, I have the following WiX variables that are all related to the binary source path in the WXS file:

1. `CargoTargetDir`
2. `CargoTarget`
3. `CargoTargetExplicit`
4. `CargoTargetPathName`
5. `Profile` (Probably should be renamed to `CargoProfile` to be consistent)

These get joined together in the printing/initialization as:

`$(var.CargoTargetDir)\$(var.CargoTargetPathName)\$(var.Profile)\Example.exe`

and `CargoTarget` does not exist in the path. However, in the preprocessor, the `CargoTarget` is used for the `CargoTargetPathName` if the `CargoTargetExplicit` variable is defined; otherwise, an empty path is used. Confusingly, a WiX variable is used as the value for the `CargoTargetPathName`. Can this preprocessor logical be moved to the Rust implementation and have a single `$(var.TargetBinaryDirectory)` or similar name variable? Then,

```rust
.arg(format!("-dTargetBinaryDirectory={}", target_binary_directory));
```

where `target_binary_directory` is determined with a method earlier in the `run` method. Maybe this conflicts with the _init_ and _print_ subcommands? 

The printed WXS template would be:

```xml
Source="$(var.TargetBinaryDirectory)\Example.exe"
```

where the `Example.exe` is set during the init/print. A CLI override could be added for the `TargetBinaryDirectory` for greater flexibility. I just don't know if the individual path components need to be separate WiX variables. I don't think you are able to change these path components when building a project in Cargo. Is this assumption correct?

A little background because I could not picture the cross compilation process: for each target, i686 and x86_64, Cargo creates a separate target directory, but only if the `--target` option is used; otherwise, it defaults to a profile folder. So, the binary source paths can be:

1. `target\debug\Example.exe`
2. `target\release\Example.exe`
3. `target\i686-pc-windows-msvc\debug\Example.exe`
4. `target\i686-pc-windows-msvc\release\Example.exe`
5. `target\x86_64-pc-windows-msvc\debug\Example.exe`
6. `target\x86_64-pc-windows-msvc\release\Example.exe`

assuming the target directory from the `cargo metadata` is `target`.

roblabla replied:
> A little background because I could not picture the cross compilation process: for each target, i686 and x86_64, Cargo creates a separate target directory, but only if the --target option is used; otherwise, it defaults to a profile folder. So, the binary source paths can be:

Yes, this is correct. I probably could move the whole logic into Rust code. The reasoning behind the current system is that I want to pass CargoTarget to the WXS so that it can be used independently. For instance, CargoTarget could be used to discriminate between -gnu and -msvc toolchains from the WXS, allowing the user to change the title or chose a different set of DLL dependencies to install.

So with CargoTarget being passed, I figured I could generate the full target path in a dynamic way. But you're probably right that instead of making the WXS generate the full path to the generated binary, it'd make more sense to let the Rust side generate it. It'd be more future-proof too, if we need to make further changes to resolve where exes are located without breaking user's code.

Here's what I could do: I'll keep CargoTarget, CargoTargetDir and Profile, as those are useful primitives for the user to use to customize their installers. I'll remove CargoTargetExplicit, and add a new CargoTargetBinDir that the rust code will generate to point to `CargoTargetDir\(CargoTarget\)?Profile\`. This way, the full set of variables sent to the WXS will be:

- Profile: The name of the Cargo profile used to build the binary. Either debug or release.
- Version: Version of the program we're installing.
- Platform (deprecated): An alias for sys.BUILDARCH, One of: x86, x64, arm, arm64
- CargoTargetDir: The path to the Cargo Target directory, equivalent to `target-dir` cargo config.
- CargoTarget: The name of the cargo target used to build the binary.
- CargoTargetBinDir: The full path where cargo will store the final binaries.

Those all seem to be independently useful. What do you think? (Also, I think having ^ documented as comments in the WXS template could be really useful for people writing the WXS ^^).

volks73 replied:
>Yes, this is correct. I probably could move the whole logic into Rust code. The reasoning behind the current system is that I want to pass CargoTarget to the WXS so that it can be used independently. For instance, CargoTarget could be used to discriminate between -gnu and -msvc toolchains from the WXS, allowing the user to change the title or chose a different set of DLL dependencies to install.

Ah, I am now getting it! The embedded WXS template may not use these variables, but a user may want access to this information in his or her own customized WXS file. Basically, cargo-wix should be providing Cargo and project build-related variables as WiX variables, like the [Cargo environment variables](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables) provided to [crates](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates) and [build scripts](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts). Got it! This is totally cool and inline with both the cargo-wix subcommand and the Cargo build process. Love it! 

This is actually a separate feature than implementing cross-compilation. I have created #118 to track and discuss a more complete implementation. We can use this PR with its initial variables as a foundation to expand out the feature. 

>Those all seem to be independently useful. What do you think? If that works for you, I'll update this PR asap.

Yes. Please implement moving the logic into Rust and the mentioned variables. I think the `Profile` variable will eventually need to be renamed to `CargoProfile` for consistency and depending on the naming convention. However, this might be a change similar to removing the `Platform` variable. I have created #121 to track deprecation and eventual removal of the `Platform` variable, and similarly I have created #122 to track renaming the `Profile` variable.

>(Also, I think having ^ documented as comments in the WXS template could be really useful for people writing the WXS).

Yes! Some comments at the top of the WXS template about the WiX variables that are provided automatically by cargo-wix should be added. A `WiX Variables` section/heading should also be added to the online help documentation, too. I have created #119 and #120 to track these separately from this PR.

I wonder if some kind of "show variables" subcommand/argument/flag/option would be useful that also lists all of the variables provided? Basically, some way to add short description of each cargo-wix-provided WiX variable like the `-h,--help` content for the CLI?

roblabla replied:
I moved CargoTargetBinDir resolution to the rust side.

volks73 reviewed (commented):
Again, thank you for implementing this feature and the effort. I saw that the `--build-plan` option might be removed and I got really confused by the `CargoTargetExplicit` implementation. I think it might be possible to move that logic back into Rust and avoid using the preprocessor. Please see my comment/rambling about it.

Once the comments are acknowledged, I am happy to merge/continue refining this implementation.

roblabla commented on src/create.rs:
```diff
@@ -902,6 +918,18 @@ impl Execution {
 
+    /// Gets the target used to compile the binary. This will either return the
+    /// target passed as an argument, or find it from rustc's HOST triple.
+    // NOTE: This does not support default-target. Ideally we would use cargo
+    // --build-plan to figure this out without having to second-guess the
+    // compiler. Unfortunately, cargo build-plan is unstable.
```
Cargo build-plan is going away, yes, but it's going to get replaced by something similar enough, like [--unit-graph](https://github.com/rust-lang/cargo/issues/8002), which contains a `platform` field.

roblabla commented on src/create.rs:
```diff
@@ -450,12 +465,17 @@ impl Execution {
         compiler
             .arg(format!("-dVersion={}", version))
-            .arg(format!("-dPlatform={}", platform))
+            .arg(format!("-dPlatform={}", platform.wix_arch()?))
```
Problem is, the warn will show up for everyone all the time, not just to users of the Platform variable.

We could grep for `var.Platform` in the provided WXS as a workaround, and show the WARN only if it's found there?

volks73 replied:
Is grep available on Windows? Could the XPath crate used in the integration tests to search usage?

I briefly looked into this after commenting and had the same conclusion/thought. While a WARN statement would possibly be a nice active notice of deprecation, I am just not sure there is a good way to do without annoying every user. 

This does highlight that a general deprecation mechanism/process is probably needed. How do other Cargo subcommands handle this? We need to be able to make changes and evolve without fear while minimizing technical debt but without constantly breaking things for everyone. The goal of a tool is to make less work for everyone. Maybe when a command is executed, like the _create_ (default) command, WARN statements with all deprecations are printed, but we add a `--surpress-deprecations` flag? This would hide the notices, or it could work in the reverse (`--show-deprecations`) and be silent by default? Hmm, maybe this could be tied to the `--nocapture` flag, such that if the output is _not_ captured, the deprecation notices are printed along with the output from the WiX applications? 

I created #123 to discuss and track this idea.

For now, the course of action will be to document everything and give users some time to migrate. I created #121 to explicitly track this and provide a more visible indication of when it is removed.

roblabla replied:
> Is grep available on Windows?

Well, by grepping, I meant looking for the string in the file, not literally invoking grep haha. XPath could certainly be used, I'll look into it.

roblabla replied:
I think this would best be resolved in a separate PR, tracked by #121. I have a potential idea of how to do this very simply, but I'll need to read a bit on how WiX works with its preprocessor first, and it should be reviewed on its own.

roblabla commented on src/print/wxs.rs:
```diff
@@ -477,6 +477,7 @@ impl Execution {
     fn default_binary_path(name: &str) -> String {
         // TODO: Handle cross-compilation?
```
Ah yes, forgot to remove it haha
